packages/melange.js/Js.ml

```diff
@@ -757,7 +757,30 @@ end = struct
     if start_idx >= end_idx then ""
     else Stdlib.String.sub str start_idx (end_idx - start_idx)
 
-  let split ?sep ?limit _str = notImplemented "Js.String" "split"
+  let split ?sep ?limit str =
+    let sep = Option.value sep ~default:str in
+    let regexp = Str.regexp_string sep in
+    let str_start_with_sep = Stdlib.String.starts_with ~prefix:sep str in
+    let str_end_with_sep = Stdlib.String.ends_with ~suffix:sep str in
+    let items =
+      (* Melange (JS) has some specific cases for splitting strings: *)
+      (* https://melange.re/unstable/playground/?language=OCaml&code=SnMubG9nKEpzLlN0cmluZy5zcGxpdCB%2Bc2VwOiJzdGFydCIgInN0YXJ0Lm9jYW1sLnJlYXNvbi5yZWFjdCIpOwpKcy5sb2coSnMuU3RyaW5nLnNwbGl0IH5zZXA6ImVuZCIgIm9jYW1sLnJlYXNvbi5yZWFjdC5lbmQiKTsKSnMubG9nKEpzLlN0cmluZy5zcGxpdCB%2Bc2VwOiJib3RoIiAiYm90aC5vY2FtbC5yZWFzb24ucmVhY3QuYm90aCIpOw%3D%3D&live=off *)
+      match (str_start_with_sep, str_end_with_sep, sep <> "") with
+      (* If the string starts and ends with the separator, we need to add an empty string at the beginning and the end *)
```

Comment:
That makes sense, right? I wouldn't not say it's a edge case

```
splitting "." "1.2.3" -> ["1", "2", "3"]
splitting "1" "1.2.3" -> ["", ".2.3"]
```

When you split you always have an empty string if it matches at the start, since it's empty.

I wonder if the `str_start_with_sep` and `str_end_with_sep` are needed

Reply:
> When you split you always have an empty string if it matches at the start, since it's empty.

Not on ocaml [Str.split](https://ocaml.org/manual/5.2/api/Str.html#VALsplit)
> An occurrence of the delimiter at the beginning or at the end of the string is ignored.

With this code:
```ocaml
  let split ?sep ?limit str =
    let sep = Option.value sep ~default:str in
    let regexp = Str.regexp_string sep in
    let items = Str.split regexp str |> Stdlib.Array.of_list in
    let limit = Option.value limit ~default:(Stdlib.Array.length items) in
    match limit with
    | limit when limit >= 0 && limit < Stdlib.Array.length items ->
        Stdlib.Array.sub items 0 limit
    | _ -> items
```

Running this test:
https://github.com/ml-in-barcelona/server-reason-react/blob/11d6afce76b24ec8d8dd9f633a9d0a2717ff9897/packages/melange.js/test.ml#L317-L319

This is the output:
```
FAIL should be equal

   Expected: `[|""; "-sep-equals-to-beginning"|]'
   Received: `[|"-sep-equals-to-beginning"|]'
```

Reply:
Right, I see on the docs they mention: `An occurrence of the delimiter at the beginning or at the end of the string is ignored.` on Str.split: https://ocaml.org/manual/5.2/api/Str.html#VALsplit_delim:~:text=An%20occurrence%20of%20the%20delimiter%20at%20the%20beginning%20or%20at%20the%20end%20of%20the%20string%20is%20ignored.

But split_delim keeps the occurrences: https://ocaml.org/manual/5.2/api/Str.html#VALsplit_delim Do you think is worth changing?

Reply:
Thank you @davesnx, I didn't know `split_delim`. It totally worth add it instead of having this complex logic.

I just had to change one thing, with `split_delim` this happened:
```ocaml
  let split ?sep ?limit str =
    let sep = Option.value sep ~default:str in
    let regexp = Str.regexp_string sep in
    let items = Str.split_delim regexp str |> Stdlib.Array.of_list in
    let limit = Option.value limit ~default:(Stdlib.Array.length items) in
    match limit with
    | limit when limit >= 0 && limit < Stdlib.Array.length items ->
        Stdlib.Array.sub items 0 limit
    | _ -> items

(* test.ml *)
assert_string_array
  (Js.String.split ~sep:"" "with-empty-sep")
  [|
    "w"; "i"; "t"; "h"; "-"; "e"; "m"; "p"; "t"; "y"; "-"; "s"; "e"; "p";
  |];
```

```
FAIL should be equal

   Expected: `[|"w"; "i"; "t"; "h"; "-"; "e"; "m"; "p"; "t"; "y"; "-"; 
                "s"; "e"; "p"|]'

   Received: `[|"w"; "i"; "t"; "h"; "-"; "e"; "m"; "p"; "t"; "y"; "-"; 
                "s"; "e"; "p"; ""|]'
```

But with a empty `sep` in Js, [it doesn't add an empty string on the end](https://github.com/ml-in-barcelona/server-reason-react/pull/168#:~:text=When%20sep%20has%20no%20value%2C%20it%20returns%20all%20chars%20splitted%20(playground)). So I was thinking on do this:
```diff

  let split ?sep ?limit str =
    let sep = Option.value sep ~default:str in
    let regexp = Str.regexp_string sep in
-   let items = Str.split_delim regexp str |> Stdlib.Array.of_list in
+   let split = if sep <> "" then Str.split_delim else Str.split in
+   let items = split regexp str |> Stdlib.Array.of_list in
    let limit = Option.value limit ~default:(Stdlib.Array.length items) in
    match limit with
    | limit when limit >= 0 && limit < Stdlib.Array.length items ->
        Stdlib.Array.sub items 0 limit
    | _ -> items
```

WDYT?

Reply:
That make sense! Thanks for the back and forth. gj